services/alertmanager/alertmanager.go

```diff
@@ -157,6 +169,159 @@ func (svc *Service) updateConfiguration(ctx context.Context) {
+		cfg.Global.SMTPAuthUsername = settings.IntegratedAlerting.EmailAlertingSettings.Username
+		cfg.Global.SMTPAuthPassword = settings.IntegratedAlerting.EmailAlertingSettings.Password
+		cfg.Global.SMTPAuthSecret = settings.IntegratedAlerting.EmailAlertingSettings.Secret
+
+		host, port, err := net.SplitHostPort(settings.IntegratedAlerting.EmailAlertingSettings.Smarthost)
+		if err != nil {
+			return errors.Errorf("Failed to set global email settings: %s", err)
+		}
+		cfg.Global.SMTPSmarthost.Host = host
+		cfg.Global.SMTPSmarthost.Port = port
+	}
+
+	if settings.IntegratedAlerting.SlackAlertingSettings != nil {
+		cfg.Global.SlackAPIURL = settings.IntegratedAlerting.SlackAlertingSettings.URL
+	}
+
+	chanMap := make(map[string]*models.Channel, len(channels))
+	for _, ch := range channels {
+		chanMap[ch.ID] = ch
+	}
+
+	recvSet := make(map[string]struct{}) // stores unique combinations of channel IDs
+	for _, r := range rules {
+		match, _ := r.GetCustomLabels()
+		match["rule_id"] = r.ID
+		// make sure same slice with different order are not considered unique.
+		sort.Strings(r.ChannelIDs)
+		recv := strings.Join(r.ChannelIDs, receiverNameSeparator)
+		recvSet[recv] = struct{}{}
+		cfg.Route.Routes = append(cfg.Route.Routes, &alertmanager.Route{
+			Match:          match,
+			Receiver:       recv,
+			RepeatInterval: promconfig.Duration(r.For),
+		})
+	}
+
+	receivers, err := generateReceivers(chanMap, recvSet)
+	if err != nil {
+		return err
+	}
+
+	cfg.Receivers = append(cfg.Receivers, receivers...)
+	return nil
+}
+
+// generateReceivers takes the channel map and a unique set of rule combinations and generates a slice of receivers.
+func generateReceivers(chanMap map[string]*models.Channel, recvSet map[string]struct{}) ([]*alertmanager.Receiver, error) {
+	receivers := make([]*alertmanager.Receiver, 0, len(recvSet))
+	for k := range recvSet {
+		recv := &alertmanager.Receiver{
+			Name: k,
+		}
+
+		individualChannels := strings.Split(k, receiverNameSeparator)
+		for _, ch := range individualChannels {
+			channel := chanMap[ch]
+			switch channel.Type {
+			case models.Email:
+				recv.EmailConfigs = append(recv.EmailConfigs, &alertmanager.EmailConfig{
+					NotifierConfig: alertmanager.NotifierConfig{
+						SendResolved: channel.EmailConfig.SendResolved,
+					},
+					To: channel.EmailConfig.To[0],
+				})
+			case models.PagerDuty:
+				pdConfig := channel.PagerDutyConfig
+				if pdConfig.RoutingKey != "" {
+					recv.PagerdutyConfigs = append(recv.PagerdutyConfigs, &alertmanager.PagerdutyConfig{
+						NotifierConfig: alertmanager.NotifierConfig{
+							SendResolved: pdConfig.SendResolved,
+						},
+						RoutingKey: pdConfig.RoutingKey,
+					})
+					break
+				}
+				recv.PagerdutyConfigs = append(recv.PagerdutyConfigs, &alertmanager.PagerdutyConfig{
+					NotifierConfig: alertmanager.NotifierConfig{
+						SendResolved: pdConfig.SendResolved,
+					},
+					ServiceKey: pdConfig.ServiceKey,
+				})
```

Comment:
`ServiceKey` also can be empty. To be more precise, only one can be specified, `RoutingKey` or `ServiceKey`.